dowhy/gcm/falsify.py

```diff
@@ -0,0 +1,885 @@
+        keys_to_rem = set()
+        for k, v in self.p_values.items():
+            if v == placeholder_val:
+                keys_to_rem.add(k)
+        for k in keys_to_rem:
+            del self.p_values[k]
+
+    def __contains__(self, item: Tuple[Union[Set, List, str], ...]) -> bool:
+        X, Y = (_to_frozenset(i) for i in item[:2])
+        if len(item) == 2 or item[2] is None:
+            Z = frozenset()
+        else:
+            Z = _to_frozenset(item[2])
+        return (X, Y, Z) in self.p_values or (Y, X, Z) in self.p_values
+
+
+def _set2list(x):
+    if isinstance(x, set):
+        return list(x)
+    return x
+
+
+def _compute_p_value(
+    data: pd.DataFrame,
+    X: Union[Set, List, str],
+    Y: Union[Set, List, str],
+    Z: Optional[Union[Set, List, str]] = None,
+    independence_test: Callable[[np.ndarray, np.ndarray], float] = kernel_based,
+    conditional_independence_test: Callable[[np.ndarray, np.ndarray, np.ndarray], float] = kernel_based,
+) -> float:
+    """Perform (conditional) independence test and report p-value.
+
+    :param data: Observations of variables in the DAG.
+    :param X: Variable to test (conditional) independence with Y
+    :param Y: Variable to test (conditional) independence with X
+    :param Z: Set to condition independence test on. Can be empty (None, empty set, or empty list).
+    :param independence_test: independence test to use.
+    :param conditional_independence_test: Conditional independence test to use.
+    :return: p-value
+    """
+
+    X = _set2list(X)
+    Y = _set2list(Y)
+    Z = _set2list(Z)
+
+    if Z:
+        p_value = conditional_independence_test(data[X].values, data[Y].values, data[Z].values)
+    else:
+        p_value = independence_test(data[X].values, data[Y].values)
+    return p_value
+
+
+def _get_parental_triples(causal_graph: DirectedGraph, include_unconditional: bool):
+    """
+    For a given graph collect all parental triples, that is, the triple (X, Y, Z) is a parental triple iff
+        Y is non-descendant of X, and
+        Z are the parents of X (can be empty if include_unconditional=True)
+    """
+    triples = []
+    for node in causal_graph.nodes:
+        parents = get_ordered_predecessors(causal_graph, node)
+        non_descendants = _get_non_descendants(causal_graph, node, exclude_parents=True)
+        if (parents or include_unconditional) and non_descendants:
+            for non_desc in non_descendants:
+                triples.append((node, non_desc, parents))
+    return triples
+
+
+def validate_lmc(
+    causal_graph: DirectedGraph,
+    data: pd.DataFrame,
+    p_values_memory: _PValuesMemory,
+    independence_test: Callable[[np.ndarray, np.ndarray], float] = kernel_based,
+    conditional_independence_test: Callable[[np.ndarray, np.ndarray, np.ndarray], float] = kernel_based,
+    significance_level: float = 0.05,
+    include_unconditional: bool = True,
+    n_jobs: int = -1,
+    **kwargs,
```

Comment:
The `kwargs` are used here because we may want to pass additional arguments to some validation method. E.g. when wer run a permutation test (`_permutation_based`) we want to also pass the reference (unpermuted) graph to `validate_parental_dsep` via `'validate_graph`. However, this is not used by the other validation methods (e.g. `validate_lmc`). `kwargs` absorbs these arguments in such cases. Is there a better solution to this?

Reply:
But the `kwargs` are not used in the function at all all (i.e., it is not passed to anything). The same applies to other functions as well. So, if the `kwargs` are not passed further, then we can just remove them from the parameter list. I only see `validate_graph` that uses the `kwargs` by passing it to another method.

Reply:
Say we want do use `validate_graph` on some graph `G_prime` that is a permutation of `G` (as done in `_permutation_based`). For this we call:
```
validate_graph(
    causal_graph=G_prime,
    data=data,
    causal_graph_reference=G,
    methods=(validate_lmc, validate_parental_dsep),
)
```
In `validate_graph` the `kwargs` are passed to the methods, i.e. `validate_lmc` and `validate_parental_dsep`. But if `validate_lmc` doesn't accept any `kwargs` even though they are not used internally, we would get: `TypeError: validate_lmc() got an unexpected keyword argument 'causal_graph_reference'`. Of course we could set an argument `causal_graph_reference=None` for `validate_lmc`(and likewise `validate_pd` and `validate_causal_minimality`) but I thought this would be very confusing. The same logic applies to why `validate_parental_dsep` requires `kwargs` (to absorb e.g. `data` when called via `validate_graph`).

Reply:
Ah, ok I think I got the issue here. First I thought it was something like:
```
def method1():
    print("a")


def method2(**kwargs):
    method1(**kwargs)
```
Which works as long as no parameter is actually passed to `method2`. But here the issue is that the parameters for `validate_lmc` and `validate_parental_dsep` are/can be different, so the passed arguments in `kwargs` would need to be absorbed and ignored in the methods where they have no usage? This can be quite confusing then. If non-default parameters should be used in certain methods, then what about using `partial` to 'pre-configure' them, i.e.:

```
from functools import partial


def method1(my_optional_parameter=None):
    print("a")

    if my_optional_parameter is not None:
        print(my_optional_parameter)

partial(method1, my_optional_parameter="Test")()
```
`partial` here makes it a callable with the specific parameters pre-set (alternatively, one can also just use `lambda`). This is also the way how we do it with e.g. the independence tests when they need to be configured. I definitely see that it is more complicated for users, but probably better than introducing `kwargs` just to ignore unknown parameters.

So here it would be e.g.:
```
validate_graph(
    causal_graph=G_prime,
    data=data,
    causal_graph_reference=G,
    methods=(partial(validate_lmc, significance_level=0.1),
             partial(validate_parental_dsep, causal_graph_reference=some_graph)),
)
```

What do you think?